codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java

```diff
@@ -292,8 +294,18 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
                 }
 
                 int size = out.size();
-                firedChannelRead |= out.insertSinceRecycled();
-                fireChannelRead(ctx, out, size);
+                if (size == 0) {
+                    decodeWasNull = true;
+                } else if (size == 1) {
+                    ctx.fireChannelRead(out.get(0));
+                } else {
+                    ArrayList<Object> ret = new ArrayList<Object>(size);
```

Comment:
It would probably be better to return the `out` list directly here (instead of alloc/copy to new list), even if it doesn't get subsequently recycled. Ideally the consuming handler would recycle it though would then need to know more about the type (maybe we could define a `RecyclableList` interface).

Also @normanmaurer I'm wondering why the existing `RecyclableArrayList` isn't used instead of the "custom" `CodecOutputList`. Is it to ensure that it's reused even when the Recycler is globally disabled?

Reply:
@njhill because we not need to handle the "different" thread scenario here.

Reply:
Thanks @normanmaurer ... if my suggestion above were to be followed then it would no longer be guaranteed to be the same thread though (even if in practice it generally would be).

As far as I know this is similar for much of the `Recycler` use (common case is same thread), and I know it's been suggested that might be where the meaningful benefits come from.. i.e. that the net degradation seen by some folks is due primarily to cross-thread recycling. Hopefully #9715 should help with further investigation of that!

Related tangent idea .. I wonder if it makes sense to have a `Recyclable` interface with `recycle()` method which could then for example be called by `ReferenceCountUtil#release()` when the passed object implements it but does not implement `ReferenceCounted`?